### PR TITLE
Change how Materialize reads in persisted data

### PR DESCRIPTION
**Pull request by ruchirK**
Tried to break this out into commits that all individually make sense and make the review easier

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4209)

### CODE REVIEW

File: src/dataflow/src/source/kafka.rs [hunk 1]
```diff
@@ -340,38 +342,26 @@ impl SourceInfo<Vec<u8>> for KafkaSourceInfo {
...
-                        e
-                    );
-                    vec![]
-                });
-                RecordIter { data, offset: 0 }.map(|r| (r.key, r.value, r.timestamp, r.offset))
-            })
-            .collect()
+    fn read_persisted_file(&mut self) -> Option<Vec<(Vec<u8>, Vec<u8>, Timestamp, i64)>> {
+        if let Some(f) = &self.persisted_files.pop() {
+            log::info!("reading data from {}", f.display());
```
Comment (ruchirK) [inline src/dataflow/src/source/kafka.rs:347]: change this to a debug log

**Review by justinj — COMMENTED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, 2 of 3 files at r4, 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4209#-:-MHHVOKK5gQH96kjtMwF:bnfp4nl)* status: 3 of 4 files reviewed, 3 unresolved discussions (waiting on @justinj and @ruchirK)

___
*[src/dataflow/src/source/kafka.rs, line 460 at r1](https://reviewable.io/reviews/materializeinc/materialize/4209#-MHHTWY7EmNvu8j83uZf:-MHHTWY80CNwlD4ndRq3:b-bv5f7a) ([raw file](https://github.com/materializeinc/materialize/blob/8cf6d28d081d5515a9b1f0230cd3fe3a7b07b95d/src/dataflow/src/source/kafka.rs#L460)):*
> ```Rust
>                     _ => unreachable!(),
>                 });
>                 // Reverse the list so we can pop elements off it in sorted order
> ```

Can we just sort it by `-meta.start_offset` instead?
___
*[src/dataflow/src/source/kafka.rs, line 345 at r5](https://reviewable.io/reviews/materializeinc/materialize/4209#-MHHVFdY91X64DXJ5M1W:-MHHVFdY91X64DXJ5M1X:b-txd6xk) ([raw file](https://github.com/materializeinc/materialize/blob/b76e57a76319c2545455d835c191157c86c56bb2/src/dataflow/src/source/kafka.rs#L345)):*
> ```Rust
>     }
> 
>     fn read_persisted_file(&mut self) -> Option<Vec<(Vec<u8>, Vec<u8>, Timestamp, i64)>> {
> ```

with this behaviour this might be better named `next_persisted_file`?

**Review by ruchirK — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4209)* status: 1 of 5 files reviewed, 3 unresolved discussions (waiting on @justinj)

___
*[src/dataflow/src/source/kafka.rs, line 460 at r1](https://reviewable.io/reviews/materializeinc/materialize/4209#-MHHTWY7EmNvu8j83uZf:-MHHe_nx3inK6zr0s4J4:b-896fix) ([raw file](https://github.com/materializeinc/materialize/blob/8cf6d28d081d5515a9b1f0230cd3fe3a7b07b95d/src/dataflow/src/source/kafka.rs#L460)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Can we just sort it by `-meta.start_offset` instead?
</blockquote></details>

Done.
___
*[src/dataflow/src/source/kafka.rs, line 345 at r5](https://reviewable.io/reviews/materializeinc/materialize/4209#-MHHVFdY91X64DXJ5M1W:-MHHeZVZ1YWM6_d1tGUy:b-896fix) ([raw file](https://github.com/materializeinc/materialize/blob/b76e57a76319c2545455d835c191157c86c56bb2/src/dataflow/src/source/kafka.rs#L345)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

with this behaviour this might be better named `next_persisted_file`?
</blockquote></details>

Done.

**Review by ruchirK — COMMENTED** [review]
I added one more commit to tweak what the generated file names look like and responded to your comments. thanks!

> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/4209#-:-MHHecJQEACb-F6SLm2Y:b-fd2ndk)* status: 1 of 5 files reviewed, 3 unresolved discussions (waiting on @justinj)

**Pull request merged**